### PR TITLE
Update to more Promises based API.

### PR DESCRIPTION
**Pull request by davidlehn**

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  env: {
```
Comment (mattcollier) [inline .eslintrc.js:3]: I think that `root: true` flag is important which is in many other places.

File: lib/main.js [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+// to keep track of listeners and the maximum number of allowed listeners.
+export function EventEmitter() {
+  this._eventListeners = new Map();
+  this._onceListeners = new Map();
+  this._maxListeners = 10;
+  this.on('maxListenersPassed', printMaxListenersWarning.bind(this));
+}
+
+// ## emitSync
+// Operates just like ``emit`` except all listeners must be synchronous.
```
Comment (mattcollier) [inline lib/main.js:33]: What's with all the backticks?

File: lib/main.js [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    // check for async result
+    if(isPromise(result)) {
+      throw new TypeError(`async listeners not allowed for "${eventName}"`);
+    }
+    // check if event canceled
+    if(result === false) {
+      return false;
+    }
+  }
+};
```
Comment (mattcollier) [inline lib/main.js:91]: It appears this returns `undefined` on success and `false` if canceled?  Is that proper?

File: lib/main.js [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+// Takes and event name and a variable number of arguments. Listeners can be
+// asynchronous and return a Promise or synchronous. Listeners can cancel
+// events if their return value is `false` or is a Promise that resolves to
+// `false`. Event cancelation immediately ends all processing of event
+// listeners and returns `false`. If the event was emitted to all listeners
+// then undefined is returned. The order of event listeners is the order they
+// were attached, but ``once`` listeners are given priority over ``on``
+// listeners (so its possible to do a one-time cancelation and not have any of
+// the ``on`` listeners called).
+EventEmitter.prototype.emit = async function emit(eventName, ...args) {
```
Comment (mattcollier) [inline lib/main.js:103]: Is there any good reason to consider moving to a class?

**Review by mattcollier — APPROVED** [review]
Thanks!

Is there any idea that this would be contributed back to the original project?  If not, should we properly break from that project and license this under BSD or something?

File: lib/main.js (hunk 2 above)
Comment (davidlehn) [reply lib/main.js:33]: Sticking with the original style. Some people prefer that so single ticks are easier to type or something. Can change if people want. This should all be converted to jsdoc too.

File: lib/main.js (hunk 3 above)
Comment (davidlehn) [reply lib/main.js:91]: It's intended.  I'm not sure if it's proper. ;-)  It's basically what the API was doing before.  I was debating if we should really break the API and return something verbose like `{cancelled: true/false}` instead. This false/undefined thing is hard to understand.

File: lib/main.js (hunk 4 above)
Comment (davidlehn) [reply lib/main.js:103]: Yes, it should be converted to a class.  I was just working with getting the code running as is with newer semantics first.  Figured that could be put off until later if needed.

**Comment by davidlehn** [comment]
Re original project: I think we took this on as the new maintainers.  I'll update something to be more clear about that.  It's MIT now, which is probably good enough?

File: lib/main.js [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  if(!this._onceListeners.has(eventName) &&
+    !this._eventListeners.has(eventName)) {
+    return;
+  }
+
+  // Copy listeners arrays to ensure listener call order
+  const onceListeners = [...(this._onceListeners.get(eventName) || [])];
+  const eventListeners = [...(this._eventListeners.get(eventName) || [])];
+
+  // process once listeners first
```
Comment (dlongley) [inline lib/main.js:45]: Listeners should really just be processed in order, independent of `once` status. I can see this biting a user later on. Don't need to change in this version if it's a headache.

Comment (dlongley) [reply lib/main.js:45]: Since it's a breaking change we can push to another version.

**Review by dlongley — APPROVED** [review]
We should update the copyright to add DB and change the note about all credit going to the previous author to say all credit for the original implementation goes to that author.

File: lib/main.js (hunk 5 above)
Comment (davidlehn) [reply lib/main.js:45]: Yeah, it's an odd choice. Just went with the original code.

File: lib/main.js (hunk 2 above)
Comment (davidlehn) [reply lib/main.js:33]: Tracking this for future release: https://github.com/digitalbazaar/async-node-events/issues/3

File: lib/main.js (hunk 3 above)
Comment (davidlehn) [reply lib/main.js:91]: Tracking for future release: https://github.com/digitalbazaar/async-node-events/issues/5

File: lib/main.js (hunk 4 above)
Comment (davidlehn) [reply lib/main.js:103]: Tracking for future: https://github.com/digitalbazaar/async-node-events/issues/7

File: lib/main.js (hunk 5 above)
Comment (davidlehn) [reply lib/main.js:45]: Tracking for future: https://github.com/digitalbazaar/async-node-events/issues/4

**Pull request merged**